src/protocols/secure_channel/CASESession.cpp

```diff
@@ -224,12 +233,29 @@ class CASESession::WorkHelper
         bool cancel = false;
         // Execute callback in background thread; data must be OK with this
         helper->mStatus = helper->mWorkCallback(helper->mData, cancel);
+
+        // Check the status and call OnWorkFailedCallback
+        if (helper->mStatus != CHIP_NO_ERROR)
+        {
+            if (auto * session = helper->mSession.load())
+            {
+                (session->*(helper->mOnWorkFailedCallback))(helper->mStatus);
+            }
+            return;
+        }
+
         VerifyOrReturn(!cancel && !helper->IsCancelled());
```

Comment:
The order of the new code and this line means that if an error is return during normal background work, and also the background work (or foreground) requests cancellation, we treat that as error instead of cancellation. I'd prefer cancellation to have precedence. (But as noted, I'd prefer normal work errors to not use this special scheduling failure callback.)

Reply:
As I removed the above checks, this comment is not valid anymore.